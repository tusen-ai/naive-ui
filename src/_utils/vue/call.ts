type AnyFunction = (...args: any[]) => any

function call (funcs: MaybeArray<() => void>): void
function call<A1> (funcs: MaybeArray<(a1: A1) => void>, a1: A1): void
function call<A1, A2> (
  funcs: MaybeArray<(a1: A1, a2: A2) => void>,
  a1: A1,
  a2: A2
): void
function call<A1, A2, A3> (
  funcs: MaybeArray<(a1: A1, a2: A2, a3: A3) => void>,
  a1: A1,
  a2: A2,
  a3: A3
): void
function call<A1, A2, A3, A4> (
  funcs: MaybeArray<(a1: A1, a2: A2, a3: A3, a4: A4) => void>,
  a1: A1,
  a2: A2,
  a3: A3,
  a4: A4
): void
function call<A extends any[]> (
  funcs: AnyFunction[] | AnyFunction,
  ...args: A
): void {
  if (Array.isArray(funcs)) {
    funcs.forEach((func) => (call as any)(func, ...args))
  } else return funcs(...args)
}

export { call }

export type MaybeArray<T> = T | T[]
