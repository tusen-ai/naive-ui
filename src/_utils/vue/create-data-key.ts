export function createDataKey(key: string | number): string {
  return typeof key === 'string' ? `s-${key}` : `n-${key}`
}
