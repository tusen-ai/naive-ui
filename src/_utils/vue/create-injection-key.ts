import type { InjectionKey } from 'vue'

export function createInjectionKey<T>(key: string): InjectionKey<T> {
  return key as any
}
