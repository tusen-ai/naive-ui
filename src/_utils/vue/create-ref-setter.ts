import type { Ref } from 'vue'

export function createRefSetter (ref: Ref<HTMLElement | null>): any {
  return (inst: { $el: HTMLElement | null } | null) => {
    if (inst) {
      ref.value = inst.$el
    } else {
      ref.value = null
    }
  }
}
