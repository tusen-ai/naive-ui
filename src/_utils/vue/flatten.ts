import {
  Fragment,
  type VNodeChild,
  createTextVNode,
  type VNode,
  Comment
} from 'vue'

// o(n) flatten
export function flatten (
  vNodes: VNodeChild[],
  filterCommentNode: boolean = true,
  result: VNode[] = []
): VNode[] {
  vNodes.forEach((vNode) => {
    if (vNode === null) return
    if (typeof vNode !== 'object') {
      if (typeof vNode === 'string' || typeof vNode === 'number') {
        result.push(createTextVNode(String(vNode)))
      }
      return
    }
    if (Array.isArray(vNode)) {
      flatten(vNode, filterCommentNode, result)
      return
    }
    if (vNode.type === Fragment) {
      if (vNode.children === null) return
      if (Array.isArray(vNode.children)) {
        flatten(vNode.children, filterCommentNode, result)
      }
      // rawSlot
    } else if (vNode.type !== Comment) {
      result.push(vNode)
    }
  })
  return result
}
