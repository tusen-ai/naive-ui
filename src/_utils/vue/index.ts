export { getSlot } from './get-slot'
export { getVNodeChildren } from './get-v-node-children'
export { keep } from './keep'
export { omit } from './omit'
export { flatten } from './flatten'
export { call } from './call'
export { keysOf } from './keysOf'
export { render } from './render'
export { getFirstSlotVNode } from './get-first-slot-vnode'
export { createDataKey } from './create-data-key'
export { createRefSetter } from './create-ref-setter'
export { createInjectionKey } from './create-injection-key'
export {
  resolveSlot,
  resolveWrappedSlot,
  resolveSlotWithProps,
  resolveWrappedSlotWithProps,
  isSlotEmpty
} from './resolve-slot'
export type { MaybeArray } from './call'
export { mergeEventHandlers } from './merge-handlers'
export { isNodeVShowFalse } from './is-node-v-show-false'
export { Wrapper } from './wrapper'
