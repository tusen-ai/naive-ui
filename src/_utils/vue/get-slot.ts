import { type ComponentPublicInstance, type VNodeChild } from 'vue'

export function getSlot (
  instance: ComponentPublicInstance,
  slotName = 'default',
  fallback: VNodeChild[] = []
): VNodeChild[] {
  const slots = instance.$slots
  const slot = slots[slotName]
  if (slot === undefined) return fallback
  return slot()
}
