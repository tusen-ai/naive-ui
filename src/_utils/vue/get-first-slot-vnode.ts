import { type Slots, type VNode } from 'vue'
import { flatten } from './flatten'
import { warn } from '../naive'

export function getFirstSlotVNode (
  slots: Slots,
  slotName = 'default',
  props: unknown = undefined
): VNode | null {
  const slot = slots[slotName]
  if (!slot) {
    warn('getFirstSlotVNode', `slot[${slotName}] is empty`)
    return null
  }
  const slotContent = flatten(slot(props))
  // vue will normalize the slot, so slot must be an array
  if (slotContent.length === 1) {
    return slotContent[0]
  } else {
    warn('getFirstSlotVNode', `slot[${slotName}] should have exactly one child`)
    return null
  }
}
