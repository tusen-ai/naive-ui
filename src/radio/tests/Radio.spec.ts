import { h } from 'vue'
import { mount } from '@vue/test-utils'
import { NRadio, NRadioGroup } from '../index'

describe('n-radio', () => {
  it('should work with import on demand', () => {
    mount(NRadio)
  })

  it('should work with `checked` prop', async () => {
    const wrapper = mount(NRadio, { props: { checked: false } })
    expect(wrapper.find('.n-radio').classes()).not.toContain('n-radio--checked')
    await wrapper.setProps({ checked: true })
    expect(wrapper.find('.n-radio').classes()).toContain('n-radio--checked')
  })

  it('should work with `defaultChecked` prop', async () => {
    let wrapper = mount(NRadio, { props: { defaultChecked: true } })
    expect(wrapper.find('.n-radio').classes()).toContain('n-radio--checked')

    wrapper = mount(NRadio, { props: { defaultChecked: false } })
    expect(wrapper.find('.n-radio').classes()).not.toContain('n-radio--checked')
  })

  it('should work with `disabled` prop', async () => {
    const wrapper = mount(NRadio, { props: { disabled: false } })
    expect(wrapper.find('.n-radio').classes()).not.toContain(
      'n-radio--disabled'
    )
    await wrapper.setProps({ disabled: true })
    expect(wrapper.find('.n-radio').classes()).toContain('n-radio--disabled')
  })

  it('should work with `name` prop', async () => {
    const wrapper = mount(NRadio, { props: { name: 'randomName111' } })

    const radio = wrapper.find('input[type=radio]')

    expect(radio.attributes('name')).toEqual('randomName111')

    await wrapper.setProps({ name: 'randomName222' })

    expect(radio.attributes('name')).toEqual('randomName222')
  })

  it('should render default slot content', async () => {
    const wrapper = mount(NRadio, { slots: { default: 'MySlotContent' } })

    const radio = wrapper.find('.n-radio__label')

    expect(radio.text()).toContain('MySlotContent')
  })

  it('should work with `label` prop', async () => {
    const wrapper = mount(NRadio, { props: { label: 'MyRandomLabel' } })

    const radio = wrapper.find('.n-radio__label')

    expect(radio.text()).toContain('MyRandomLabel')

    await wrapper.setProps({ label: 'MyNewRandomLabel' })

    expect(radio.text()).toContain('MyNewRandomLabel')
  })

  it('should render default slot content in priority of label', async () => {
    const wrapper = mount(NRadio, {
      props: { label: 'MyRandomLabel' },
      slots: { default: 'MySlotContent' }
    })

    const radio = wrapper.find('.n-radio__label')

    expect(radio.text()).not.toContain('MyRandomLabel')
    expect(radio.text()).toContain('MySlotContent')
  })

  it('should work with `size` prop', async () => {
    ;(['small', 'medium', 'large'] as const).forEach((size) => {
      const wrapper = mount(NRadio, { props: { size: size } })
      expect(wrapper.find('.n-radio').attributes('style')).toMatchSnapshot()
    })
  })

  it('should work with `onUpdate:checked` prop', async () => {
    const onUpdate1 = vi.fn()
    const onUpdate2 = vi.fn()
    const wrapper = mount(NRadio, {
      props: { 'onUpdate:checked': onUpdate1, onUpdateChecked: onUpdate2 }
    })

    await wrapper.find('.n-radio').trigger('click')
    setTimeout(() => {
      expect(onUpdate1).toHaveBeenCalled()
      expect(onUpdate2).toHaveBeenCalled()
    }, 0)
  })
})

describe('n-radio-group', () => {
  it('should work with import on demand', () => {
    mount(NRadioGroup, {
      slots: {
        default: () => [
          h(NRadio, null, { default: () => 'test-item1' }),
          h(NRadio, null, { default: () => 'test-item2' })
        ]
      }
    })
  })

  it('should work with `disabled` prop', async () => {
    const wrapper = mount(NRadioGroup, {
      slots: {
        default: () => [
          h(NRadio, null, { default: () => 'test-item1' }),
          h(NRadio, null, { default: () => 'test-item2' })
        ]
      }
    })
    expect(wrapper.find('.n-radio--disabled').exists()).not.toBe(true)

    await wrapper.setProps({ disabled: true })
    expect(wrapper.findAll('.n-radio')[0].classes()).toContain(
      'n-radio--disabled'
    )
    expect(wrapper.findAll('.n-radio')[1].classes()).toContain(
      'n-radio--disabled'
    )
  })

  it('should work with `name` prop', async () => {
    const wrapper = mount(NRadioGroup, {
      props: {
        name: 'randomName111'
      },
      slots: {
        default: () => [
          h(NRadio, null, { default: () => 'test-item1' }),
          h(NRadio, null, { default: () => 'test-item2' })
        ]
      }
    })

    const radio1 = wrapper.findAll('input[type=radio]')[0]
    const radio2 = wrapper.findAll('input[type=radio]')[1]

    expect(radio1.attributes('name')).toEqual('randomName111')
    expect(radio2.attributes('name')).toEqual('randomName111')

    await wrapper.setProps({ name: 'randomName222' })

    expect(radio1.attributes('name')).toEqual('randomName222')
    expect(radio2.attributes('name')).toEqual('randomName222')
  })

  it('should work with `size` prop', async () => {
    ;(['small', 'medium', 'large'] as const).forEach((size) => {
      const wrapper = mount(NRadioGroup, {
        props: {
          size: size
        },
        slots: {
          default: () => [
            h(NRadio, null, { default: () => 'test-item1' }),
            h(NRadio, null, { default: () => 'test-item2' })
          ]
        }
      })
      expect(
        wrapper.find('.n-radio-group').attributes('style')
      ).toMatchSnapshot()
    })
  })

  it('should work with `value` prop', async () => {
    const wrapper = mount(NRadioGroup, {
      props: {
        value: 'test1'
      },
      slots: {
        default: () => [
          h(NRadio, { value: 'test1' }, { default: () => 'test-item1' }),
          h(NRadio, { value: 'test2' }, { default: () => 'test-item2' })
        ]
      }
    })
    expect(wrapper.findAll('.n-radio')[0].classes()).toContain(
      'n-radio--checked'
    )
    expect(wrapper.findAll('.n-radio')[1].classes()).not.toContain(
      'n-radio--checked'
    )

    await wrapper.setProps({ value: 'test2' })
    expect(wrapper.findAll('.n-radio')[1].classes()).toContain(
      'n-radio--checked'
    )
    expect(wrapper.findAll('.n-radio')[0].classes()).not.toContain(
      'n-radio--checked'
    )
  })

  it('should work with `on-update:value` prop', async () => {
    const onUpdate = vi.fn()
    const wrapper = mount(NRadioGroup, {
      props: {
        onUpdateValue: onUpdate
      },
      slots: {
        default: () => [
          h(NRadio, { value: 'test1' }, { default: () => 'test-item1' }),
          h(NRadio, { value: 'test2' }, { default: () => 'test-item2' })
        ]
      }
    })

    await wrapper.findAll('.n-radio')[1].trigger('click')
    setTimeout(() => {
      expect(onUpdate).toHaveBeenCalled()
    }, 0)
  })
})
