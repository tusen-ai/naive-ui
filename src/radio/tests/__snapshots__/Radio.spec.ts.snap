// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`n-radio should work with \`size\` prop 1`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-label-line-height: 1.6; --n-label-font-weight: 400; --n-box-shadow: inset 0 0 0 1px rgb(224, 224, 230); --n-box-shadow-active: inset 0 0 0 1px #18a058; --n-box-shadow-disabled: inset 0 0 0 1px rgb(224, 224, 230); --n-box-shadow-focus: inset 0 0 0 1px #18a058, 0 0 0 2px rgba(24, 160, 88, 0.2); --n-box-shadow-hover: inset 0 0 0 1px #18a058; --n-color: #FFF; --n-color-active: #0000; --n-color-disabled: rgb(250, 250, 252); --n-dot-color-active: #18a058; --n-dot-color-disabled: rgb(224, 224, 230); --n-font-size: 14px; --n-radio-size: 14px; --n-text-color: rgb(51, 54, 57); --n-text-color-disabled: rgba(194, 194, 194, 1); --n-label-padding: 0 8px;"`;

exports[`n-radio should work with \`size\` prop 2`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-label-line-height: 1.6; --n-label-font-weight: 400; --n-box-shadow: inset 0 0 0 1px rgb(224, 224, 230); --n-box-shadow-active: inset 0 0 0 1px #18a058; --n-box-shadow-disabled: inset 0 0 0 1px rgb(224, 224, 230); --n-box-shadow-focus: inset 0 0 0 1px #18a058, 0 0 0 2px rgba(24, 160, 88, 0.2); --n-box-shadow-hover: inset 0 0 0 1px #18a058; --n-color: #FFF; --n-color-active: #0000; --n-color-disabled: rgb(250, 250, 252); --n-dot-color-active: #18a058; --n-dot-color-disabled: rgb(224, 224, 230); --n-font-size: 14px; --n-radio-size: 16px; --n-text-color: rgb(51, 54, 57); --n-text-color-disabled: rgba(194, 194, 194, 1); --n-label-padding: 0 8px;"`;

exports[`n-radio should work with \`size\` prop 3`] = `"--n-bezier: cubic-bezier(.4, 0, .2, 1); --n-label-line-height: 1.6; --n-label-font-weight: 400; --n-box-shadow: inset 0 0 0 1px rgb(224, 224, 230); --n-box-shadow-active: inset 0 0 0 1px #18a058; --n-box-shadow-disabled: inset 0 0 0 1px rgb(224, 224, 230); --n-box-shadow-focus: inset 0 0 0 1px #18a058, 0 0 0 2px rgba(24, 160, 88, 0.2); --n-box-shadow-hover: inset 0 0 0 1px #18a058; --n-color: #FFF; --n-color-active: #0000; --n-color-disabled: rgb(250, 250, 252); --n-dot-color-active: #18a058; --n-dot-color-disabled: rgb(224, 224, 230); --n-font-size: 15px; --n-radio-size: 18px; --n-text-color: rgb(51, 54, 57); --n-text-color-disabled: rgba(194, 194, 194, 1); --n-label-padding: 0 8px;"`;

exports[`n-radio-group should work with \`size\` prop 1`] = `"--n-font-size: 14px; --n-bezier: cubic-bezier(.4, 0, .2, 1); --n-button-border-color: rgb(224, 224, 230); --n-button-border-color-active: #18a058; --n-button-border-radius: 3px; --n-button-box-shadow: inset 0 0 0 1px #0000; --n-button-box-shadow-focus: inset 0 0 0 1px #18a058, 0 0 0 2px rgba(24, 160, 88, 0.3); --n-button-box-shadow-hover: inset 0 0 0 1px #0000; --n-button-color-active: #FFF; --n-button-text-color: rgb(51, 54, 57); --n-button-text-color-hover: #18a058; --n-button-text-color-active: #18a058; --n-height: 28px; --n-opacity-disabled: 0.5;"`;

exports[`n-radio-group should work with \`size\` prop 2`] = `"--n-font-size: 14px; --n-bezier: cubic-bezier(.4, 0, .2, 1); --n-button-border-color: rgb(224, 224, 230); --n-button-border-color-active: #18a058; --n-button-border-radius: 3px; --n-button-box-shadow: inset 0 0 0 1px #0000; --n-button-box-shadow-focus: inset 0 0 0 1px #18a058, 0 0 0 2px rgba(24, 160, 88, 0.3); --n-button-box-shadow-hover: inset 0 0 0 1px #0000; --n-button-color-active: #FFF; --n-button-text-color: rgb(51, 54, 57); --n-button-text-color-hover: #18a058; --n-button-text-color-active: #18a058; --n-height: 34px; --n-opacity-disabled: 0.5;"`;

exports[`n-radio-group should work with \`size\` prop 3`] = `"--n-font-size: 15px; --n-bezier: cubic-bezier(.4, 0, .2, 1); --n-button-border-color: rgb(224, 224, 230); --n-button-border-color-active: #18a058; --n-button-border-radius: 3px; --n-button-box-shadow: inset 0 0 0 1px #0000; --n-button-box-shadow-focus: inset 0 0 0 1px #18a058, 0 0 0 2px rgba(24, 160, 88, 0.3); --n-button-box-shadow-hover: inset 0 0 0 1px #0000; --n-button-color-active: #FFF; --n-button-text-color: rgb(51, 54, 57); --n-button-text-color-hover: #18a058; --n-button-text-color-active: #18a058; --n-height: 40px; --n-opacity-disabled: 0.5;"`;
